### Main Challenge:

You must create a Python App (Data Pipeline) that allow their potential users to find the nearest BiciMAD station to a set of places of interest using the methods included in the module geo_calculations.py.

### Instalaciones accesibles no municipales

Recopila información de todas aquellas instalaciones accesibles no municipales de la ciudad de las que posee conocimiento expreso el Ayuntamiento de Madrid y que están publicadas en el portal institucional madrid.es. 

Incluye, entre otras, información sobre:

-          Ministerios

-          Oficinas de correos

-          Centros de enseñanza y universidades

-          Centros de día y residenciales

-          Museos

-          Espacios de ocio

 

En la información de cada espacio, incluye información (si está disponible) sobre:

- el horario
- los equipamientos de que dispone y/o los servicios que presta
- su dirección completa (incluyendo calle,  distrito, barrio, código postal y coordenadas geográficas)
- los medios de transporte más cercanos disponibles (Bus, metro, Bicimad y Renfe)
- teléfono
- el correo electrónico
- la dirección web donde está publicada en madrid.es para su consulta en un formato más amigable.

**[Enlace](https://datos.madrid.es/portal/site/egob/menuitem.c05c1f754a33a9fbe4b2e4b284f1a5a0/?vgnextoid=f31050abd15b3410VgnVCM1000000b205a0aRCRD&vgnextchannel=374512b9ace9f310VgnVCM100000171f5a0aRCRD&vgnextfmt=default)**

Los métodos incluidos en el módulo geo_calculations.py **mencionados anteriormente** son copiados-pegados textualmente a continuación:

In [1]:
# conda install -c conda-forge geopandas 
# (shapely is a geopandas dependance)

from shapely.geometry import Point # https://geopandas.org/en/stable/docs/reference/api/geopandas.GeoSeries.html
import geopandas as gpd # https://geopandas.org/en/stable/getting_started.html

# Methods

def to_mercator(lat, long):
    # transform latitude/longitude data in degrees to pseudo-mercator coordinates in metres
    c = gpd.GeoSeries([Point(lat, long)], crs=4326)
    c = c.to_crs(3857)
    return c

def distance_meters(lat_start, long_start, lat_finish, long_finish):
    # return the distance in metres between to latitude/longitude pair points in degrees 
    # (e.g.: Start Point -> 40.4400607 / -3.6425358 End Point -> 40.4234825 / -3.6292625)
    start = to_mercator(lat_start, long_start)
    finish = to_mercator(lat_finish, long_finish)
    return start.distance(finish)

Sin embargo, tuve problemas para instalar Geopandas. Así que tuve que hacerlo desde el Notebook con las siguientes líneas de comandos:

- `%pip install geopandas`
- `%pip install shapely`

Esta es la [documentación](https://geopandas.org/en/stable/docs/reference/geoseries.html) de **Geoseries** en Pandas.

Por supuesto, importamos las siguientes **librerías:**

In [2]:
import pandas as pd
import numpy as np

#### Lo que generamos de los notebooks anteriores, importamos las versiones de los siguientes .cvs:

- bicimad_df
- biciparks_df
- inst_no_munic_df

#### BiciMAD:

In [3]:
bicimad_df = pd.read_csv('../data/processed/bicimad_stations.csv')
bicimad_df = bicimad_df.rename(columns={"mad_long": "BiciMAD Long.",
                                        "mad_lat" : "BiciMAD Lat."})
bicimad_df = bicimad_df.drop(columns='Unnamed: 0')
bicimad_df

BiciMAD station  \
0           1a - Puerta del Sol A   
1           1b - Puerta del Sol B   
2                 2 - Miguel Moya   
3          3 - Plaza Conde Suchil   
4                    4 - Malasaña   
..                            ...   
259                    257 - INEF   
260  258 - Ciudad Universitaria 1   
261  259 - Ciudad Universitaria 2   
262       260 - Facultad Biología   
263        261 - Facultad Derecho   

                                       BiciMAD address  BiciMAD Long.  \
0                                  Puerta del Sol nº 1      -3.701834   
1                                  Puerta del Sol nº 1      -3.701603   
2                               Calle Miguel Moya nº 1      -3.705842   
3             Plaza del Conde del Valle de Súchil nº 3      -3.706917   
4                          Calle Manuela Malasaña nº 5      -3.702587   
..                                                 ...            ...   
259  Avenida Juan de Herrera frente a la calle Paul...      -3.729970   
260  Avenida de la Complutense (Metro Ciudad Univer...      -3.726990   
261  Avenida de la Complutense (Metro Ciudad Univer...      -3.726930   
262          Calle José Antonio Novais frente al nº 12      -3.727295   
263                          Avenida Complutense nº 23      -3.729370   

     BiciMAD Lat.  
0       40.417214  
1       40.417313  
2       40.420589  
3       40.430294  
4       40.428552  
..            ...  
259     40.438960  
260     40.443750  
261     40.443420  
262     40.448332  
263     40.451090  

[264 rows x 4 columns]

#### Biciparks:

In [4]:
biciparks_df = pd.read_csv('../data/processed/biciparks_df.csv')
biciparks_df = biciparks_df.rename(columns={"park_long": "BiciPARK Long.",
                                            "park_lat" : "BiciPARK Lat."})
biciparks_df = biciparks_df.drop(columns=['Unnamed: 0','zip_code'])
biciparks_df

BiciPARK station                  BiciPARK address  \
0  Bicipark Fuente de la Mora     Estación de Fuente de la Mora   
1             Bicipark Orense         Avenida General Perón, 27   
2          Bicipark Montalban             Calle de Montalbán, 5   
3            Bicipark Almagro                 Calle Almagro, 11   
4          Bicipark Recoletos             Paseo de Recoletos, 2   
5           Bicipark Recuerdo                  Calle Hiedra, 26   
6          Bicipark Salamanca      Pl. del Marqués de Salamanca   
7     Bicipark Villa de Paris         Calle de Orellana, 23, 21   
8              Bicipark Pitis  Calle Pau Arroyo del Fresno, 181   
9           Bicipark Aviación         Parking Aviación Española   

   BiciPARK Long.  BiciPARK Lat.  
0       -3.663068      40.484684  
1       -3.693897      40.452973  
2       -3.690786      40.418146  
3       -3.692879      40.430361  
4       -3.692245      40.420354  
5       -3.678921      40.472419  
6       -3.679172      40.430315  
7       -3.693698      40.425608  
8       -3.724914      40.495942  
9       -3.783695      40.383492

#### Instalaciones Accesibles NO municipales:

In [5]:
inst_no_munic_df = pd.read_csv('../data/processed/inst_no_munic_df.csv')
inst_no_munic_df = inst_no_munic_df.drop(columns='Unnamed: 0')
inst_no_munic_df = inst_no_munic_df.drop(columns='Postal code')
inst_no_munic_df = inst_no_munic_df.rename(columns={"place_lat": "Place LAT.",
                                            "place_long" : "Place LONG."})
inst_no_munic_df

Place of interest  \
0                            Sede QuedaT.com Chamartín   
1    A (mas) ECU control urbanístico, S.L. Entidad ...   
2    APFMPS, Residencia, centro de día y centro ocu...   
3    ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
4                   Administración Ciudad Lineal. AEAT   
..                                                 ...   
404                            Tanatorio Sur de Madrid   
405   Teatro Municipal de Títeres. Parque de El Retiro   
406                                  Teatros del Canal   
407                                      WiZink Center   
408  Yacimiento paleontológico. Estación de Metro C...   

                                         Place address  Place LAT.  \
0                               Pasaje Doña Carlota 13   40.447907   
1               Calle Añastro 11 Planta Baja Puerta 10   40.468046   
2                        Carretera Colmenar Viejo 13.6   40.562993   
3                               Camino Valderribas 115   40.394474   
4                                     Calle Uruguay 16   40.454894   
..                                                 ...         ...   
404                Calle Ildefonso Gonzalez Valencia 1   40.373252   
405  Avenida México 4 Parque De El Retiro. Entrada ...   40.418784   
406                               Calle Cea Bermudez 1   40.438548   
407                                Calle Jorge Juan 99   40.423273   
408                            Calle Via Carpetana 141   40.392541   

     Place LONG.  
0      -3.663171  
1      -3.663242  
2      -3.714052  
3      -3.656134  
4      -3.672454  
..           ...  
404    -3.721590  
405    -3.686652  
406    -3.705057  
407    -3.672101  
408    -3.741303  

[409 rows x 4 columns]

El output debe lucir algo ["así"](https://github.com/ih-datapt-mad/ih_datamadpt0923_project_m1)

Vamos a usar la función para **calcular distancias** entre coordenadas:

In [6]:
def distance_meters(lat_start, long_start, lat_finish, long_finish):
    # return the distance in metres between to latitude/longitude pair points in degrees 
    # (e.g.: Start Point -> 40.4400607 / -3.6425358 End Point -> 40.4234825 / -3.6292625)
    start = to_mercator(lat_start, long_start)
    finish = to_mercator(lat_finish, long_finish)
    return start.distance(finish)

Haremos una primera prueba desde **BiciMAD** 1a - Puerta del Sol A, calculando su distancia hasta **Sede QuedaT.com Chamartín**

In [7]:
bicimad_df.head(1)

BiciMAD station      BiciMAD address  BiciMAD Long.  BiciMAD Lat.
0  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214

In [8]:
inst_no_munic_df.head(1)

Place of interest           Place address  Place LAT.  Place LONG.
0  Sede QuedaT.com Chamartín  Pasaje Doña Carlota 13   40.447907    -3.663171

In [9]:
# distance_meters(lat_start, long_start, lat_finish, long_finish)

# El punto inicial es Puerta del Sol (lat_start, long_start) y el destino es Chamartín (lat_finish, long_finish)

distance_meters(40.417214,-3.701834, 40.447907,-3.663171)

0    5502.246488
dtype: float64

La primera prueba nos dice que hay una distancia de 5502 mts (5.5 kms) **desde** Bicimad de Puerta del Sol **hasta** Sede QuedaT.com Chamartín.

### Combinando Dataframes con [Pandas Cross-join](https://linuxhint.com/pandas-cross-join/)

#### 1. Entre BiciMAD y los Sitios de interés

In [10]:
# Creamos una columna 'key' en bicimad_df y en inst_no_munic_df para tener una llave en común 
bicimad_df['key'] = 0
inst_no_munic_df['key'] = 0

# Creamos un nuevo df bicimad_project_df, haciendo el merge de bicimad con el de los sitios de interés
bicimad_project_df = bicimad_df.merge(inst_no_munic_df, on='key', how='outer')
bicimad_project_df.head()

BiciMAD station      BiciMAD address  BiciMAD Long.  BiciMAD Lat.  \
0  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214   
1  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214   
2  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214   
3  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214   
4  1a - Puerta del Sol A  Puerta del Sol nº 1      -3.701834     40.417214   

   key                                  Place of interest  \
0    0                          Sede QuedaT.com Chamartín   
1    0  A (mas) ECU control urbanístico, S.L. Entidad ...   
2    0  APFMPS, Residencia, centro de día y centro ocu...   
3    0  ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
4    0                 Administración Ciudad Lineal. AEAT   

                            Place address  Place LAT.  Place LONG.  
0                  Pasaje Doña Carlota 13   40.447907    -3.663171  
1  Calle Añastro 11 Planta Baja Puerta 10   40.468046    -3.663242  
2           Carretera Colmenar Viejo 13.6   40.562993    -3.714052  
3                  Camino Valderribas 115   40.394474    -3.656134  
4                        Calle Uruguay 16   40.454894    -3.672454

In [11]:
# Combinando cada fila de las estaciones bicimad con cada una de los sitios de interés, obtenemos 107712 filas
len(bicimad_project_df)

107976

#### 2. Entre Biciparks y los Sitios de interés

In [12]:
# Creamos una columna 'key' en biciparks_df y en inst_no_munic_df para tener una llave en común 
biciparks_df['key'] = 0

# Creamos un nuevo df parks_project_df, haciendo el merge de biciparks con el de los sitios de interés
parks_project_df = biciparks_df.merge(inst_no_munic_df, on='key', how='outer')
parks_project_df.head()

BiciPARK station               BiciPARK address  BiciPARK Long.  \
0  Bicipark Fuente de la Mora  Estación de Fuente de la Mora       -3.663068   
1  Bicipark Fuente de la Mora  Estación de Fuente de la Mora       -3.663068   
2  Bicipark Fuente de la Mora  Estación de Fuente de la Mora       -3.663068   
3  Bicipark Fuente de la Mora  Estación de Fuente de la Mora       -3.663068   
4  Bicipark Fuente de la Mora  Estación de Fuente de la Mora       -3.663068   

   BiciPARK Lat.  key                                  Place of interest  \
0      40.484684    0                          Sede QuedaT.com Chamartín   
1      40.484684    0  A (mas) ECU control urbanístico, S.L. Entidad ...   
2      40.484684    0  APFMPS, Residencia, centro de día y centro ocu...   
3      40.484684    0  ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
4      40.484684    0                 Administración Ciudad Lineal. AEAT   

                            Place address  Place LAT.  Place LONG.  
0                  Pasaje Doña Carlota 13   40.447907    -3.663171  
1  Calle Añastro 11 Planta Baja Puerta 10   40.468046    -3.663242  
2           Carretera Colmenar Viejo 13.6   40.562993    -3.714052  
3                  Camino Valderribas 115   40.394474    -3.656134  
4                        Calle Uruguay 16   40.454894    -3.672454

In [13]:
# Combinando cada fila de las estaciones biciparks con cada una de los sitios de interés, obtenemos 4080 filas
len(parks_project_df)

4090

### Calculando distancias entre coordenadas con def distance_meters

#### 1. Distancia en metros desde BiciMAD de destino hasta cada Sitio de interés

In [14]:
# distance_meters(lat_start, long_start, lat_finish, long_finish)
bicimad_project_df['Distance from station in mts'] = round(bicimad_project_df.apply(lambda row: distance_meters(row['BiciMAD Lat.'], row['BiciMAD Long.'], row['Place LAT.'], row['Place LONG.']), axis=1), 2)

# Ordenando de menor a mayor distancia
bicimad_project_df = bicimad_project_df.sort_values(by='Distance from station in mts')

(La función se tomó **10 minutos** en calcular las distancias de todas las filas)

In [15]:
# Cambiando el orden de las columnas
column_order = ['Place of interest','Place address','BiciMAD station', 'BiciMAD address','Distance from station in mts']
bicimad_project_df = bicimad_project_df[column_order]
bicimad_project_df = bicimad_project_df.sort_values(by='Distance from station in mts')
bicimad_project_df

Place of interest  \
18436          Biblioteca Pública Pedro Salinas (Centro)   
2572     Colegio Oficial de Arquitectos de Madrid (COAM)   
53539        Residencia Universitaria Augustinus Nebrija   
26630                                  CaixaForum Madrid   
38057                      Biblioteca Nacional de España   
...                                                  ...   
91580       Residencia para mayores Juan XXIII (Aravaca)   
91209  APFMPS, Residencia, centro de día y centro ocu...   
91495  Oficina de Correos. Sucursal 32. Plaza Rogelio...   
91559                           Palacio Real de El Pardo   
91379                    Espacio para mayores Las Tablas   

                       Place address           BiciMAD station  \
18436    Glorieta Puerta De Toledo 1     45 - Puerta de Toledo   
2572              Calle Hortaleza 63   6 - Colegio Arquitectos   
53539          Calle Cea Bermudez 59        128 - Cea Bermúdez   
26630                 Paseo Prado 36              67 - Almadén   
38057             Paseo Recoletos 20  94 - Biblioteca Nacional   
...                              ...                       ...   
91580               Calle Procion 10             221 - Pavones   
91209  Carretera Colmenar Viejo 13.6             221 - Pavones   
91495       Plaza Rogelio Enriquez 5             221 - Pavones   
91559          Calle Manuel Alonso 8             221 - Pavones   
91379  Avenida Camino De Santiago 39             221 - Pavones   

                            BiciMAD address  Distance from station in mts  
18436  Glorieta de la Puerta de Toledo nº 1                          6.75  
2572                  Calle Hortaleza nº 63                          8.89  
53539              Calle Cea Bermúdez nº 59                         12.55  
26630                   Calle Almadén nº 28                         23.87  
38057              Paseo de Recoletos nº 20                         25.23  
...                                     ...                           ...  
91580      Calle Hacienda de Pavones nº 342                      19922.79  
91209      Calle Hacienda de Pavones nº 342                      20157.02  
91495      Calle Hacienda de Pavones nº 342                      20647.57  
91559      Calle Hacienda de Pavones nº 342                      20702.45  
91379      Calle Hacienda de Pavones nº 342                      20769.46  

[107976 rows x 5 columns]

In [16]:
bicimad_project_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 107976 entries, 18436 to 91379
Data columns (total 5 columns):
 #   Column                        Non-Null Count   Dtype  
---  ------                        --------------   -----  
 0   Place of interest             107976 non-null  object 
 1   Place address                 107976 non-null  object 
 2   BiciMAD station               107976 non-null  object 
 3   BiciMAD address               107976 non-null  object 
 4   Distance from station in mts  107976 non-null  float64
dtypes: float64(1), object(4)
memory usage: 4.9+ MB


#### 2. Distancia en metros desde Biciparks de destino hasta cada Sitio de interés

In [17]:
parks_project_df['Distance from parking in mts'] = round(parks_project_df.apply(lambda row: distance_meters(row['BiciPARK Lat.'], row['BiciPARK Long.'], row['Place LAT.'], row['Place LONG.']), axis=1), 2)
parks_project_df.sort_values(by='Distance from parking in mts')

BiciPARK station           BiciPARK address  BiciPARK Long.  \
826        Bicipark Montalban      Calle de Montalbán, 5       -3.690786   
2906  Bicipark Villa de Paris  Calle de Orellana, 23, 21       -3.693698   
1902       Bicipark Recoletos      Paseo de Recoletos, 2       -3.692245   
1084       Bicipark Montalban      Calle de Montalbán, 5       -3.690786   
2902  Bicipark Villa de Paris  Calle de Orellana, 23, 21       -3.693698   
...                       ...                        ...             ...   
3992        Bicipark Aviación  Parking Aviación Española       -3.783695   
3758        Bicipark Aviación  Parking Aviación Española       -3.783695   
3996        Bicipark Aviación  Parking Aviación Española       -3.783695   
3692        Bicipark Aviación  Parking Aviación Española       -3.783695   
3901        Bicipark Aviación  Parking Aviación Española       -3.783695   

      BiciPARK Lat.  key                                  Place of interest  \
826       40.418146    0                     Administración Montalbán. AEAT   
2906      40.425608    0                    Biblioteca del Tribunal Supremo   
1902      40.420354    0  Oficina de Correos. Oficina Principal. Paseo d...   
1084      40.418146    0  Oficina de Correos. Oficina Principal. Paseo d...   
2902      40.425608    0  Biblioteca del Consejo General del Poder judicial   
...             ...  ...                                                ...   
3992      40.383492    0      Oficina de Correos. Sucursal 59. Calle Alanis   
3758      40.383492    0  Centro de día y centro de rehabilitación Funda...   
3996      40.383492    0  Oficina de Correos. Sucursal 64. Aeropuerto T1...   
3692      40.383492    0         Aeropuerto Adolfo Suarez Madrid -  Barajas   
3901      40.383492    0  Intercambiador de transporte. Aeropuerto T1 - ...   

                                   Place address  Place LAT.  Place LONG.  \
826                            Calle Montalban 6   40.417903    -3.690676   
2906                      Plaza Villa De Paris 1   40.424744    -3.693302   
1902                               Paseo Prado 1   40.418545    -3.692583   
1084                               Paseo Prado 1   40.418545    -3.692583   
2902              Calle Marques De La Ensenada 8   40.424136    -3.692518   
...                                          ...         ...          ...   
3992                              Calle Alanis 1   40.471358    -3.583066   
3758  Plaza Hermanos Falco Y Alvarez De Toledo 1   40.473511    -3.578051   
3996      Aeropuerto T1 T2 T3  Sala 1 (Llegadas)   40.463855    -3.570391   
3692                          Avenida Hispanidad   40.468203    -3.569632   
3901                       Avenida Hispanidad 44   40.468207    -3.569629   

      Distance from parking in mts  
826                          29.74  
2906                        105.83  
1902                        204.91  
1084                        205.33  
2902                        210.18  
...                            ...  
3992                      24424.29  
3758                      25032.79  
3996                      25420.08  
3692                      25673.22  
3901                      25673.64  

[4090 rows x 10 columns]

In [18]:
# Cambiando el orden de las columnas
column_order = ['Place of interest','Place address','BiciPARK station', 'BiciPARK address','Distance from parking in mts']
parks_project_df = parks_project_df[column_order]
parks_project_df = parks_project_df.sort_values(by='Distance from parking in mts')
parks_project_df

Place of interest  \
826                      Administración Montalbán. AEAT   
2906                    Biblioteca del Tribunal Supremo   
1902  Oficina de Correos. Oficina Principal. Paseo d...   
1084  Oficina de Correos. Oficina Principal. Paseo d...   
2902  Biblioteca del Consejo General del Poder judicial   
...                                                 ...   
3992      Oficina de Correos. Sucursal 59. Calle Alanis   
3758  Centro de día y centro de rehabilitación Funda...   
3996  Oficina de Correos. Sucursal 64. Aeropuerto T1...   
3692         Aeropuerto Adolfo Suarez Madrid -  Barajas   
3901  Intercambiador de transporte. Aeropuerto T1 - ...   

                                   Place address         BiciPARK station  \
826                            Calle Montalban 6       Bicipark Montalban   
2906                      Plaza Villa De Paris 1  Bicipark Villa de Paris   
1902                               Paseo Prado 1       Bicipark Recoletos   
1084                               Paseo Prado 1       Bicipark Montalban   
2902              Calle Marques De La Ensenada 8  Bicipark Villa de Paris   
...                                          ...                      ...   
3992                              Calle Alanis 1        Bicipark Aviación   
3758  Plaza Hermanos Falco Y Alvarez De Toledo 1        Bicipark Aviación   
3996      Aeropuerto T1 T2 T3  Sala 1 (Llegadas)        Bicipark Aviación   
3692                          Avenida Hispanidad        Bicipark Aviación   
3901                       Avenida Hispanidad 44        Bicipark Aviación   

               BiciPARK address  Distance from parking in mts  
826       Calle de Montalbán, 5                         29.74  
2906  Calle de Orellana, 23, 21                        105.83  
1902      Paseo de Recoletos, 2                        204.91  
1084      Calle de Montalbán, 5                        205.33  
2902  Calle de Orellana, 23, 21                        210.18  
...                         ...                           ...  
3992  Parking Aviación Española                      24424.29  
3758  Parking Aviación Española                      25032.79  
3996  Parking Aviación Española                      25420.08  
3692  Parking Aviación Española                      25673.22  
3901  Parking Aviación Española                      25673.64  

[4090 rows x 5 columns]

In [19]:
parks_project_df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 4090 entries, 826 to 3901
Data columns (total 5 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   Place of interest             4090 non-null   object 
 1   Place address                 4090 non-null   object 
 2   BiciPARK station              4090 non-null   object 
 3   BiciPARK address              4090 non-null   object 
 4   Distance from parking in mts  4090 non-null   float64
dtypes: float64(1), object(4)
memory usage: 191.7+ KB


# Distancias cortas a través de un groupby()

#### 1. BiciMAD

In [20]:
min_bicimad_df = bicimad_project_df.groupby('Place of interest', as_index=False).agg({'Place address': 'first', 'Distance from station in mts': 'min'})
min_bicimad_df

Place of interest  \
0    A (mas) ECU control urbanístico, S.L. Entidad ...   
1    APFMPS, Residencia, centro de día y centro ocu...   
2    ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
3                   Administración Ciudad Lineal. AEAT   
4                Administración Fermín Caballero. AEAT   
..                                                 ...   
404                            Tanatorio Sur de Madrid   
405   Teatro Municipal de Títeres. Parque de El Retiro   
406                                  Teatros del Canal   
407                                      WiZink Center   
408  Yacimiento paleontológico. Estación de Metro C...   

                                         Place address  \
0               Calle Añastro 11 Planta Baja Puerta 10   
1                        Carretera Colmenar Viejo 13.6   
2                               Camino Valderribas 115   
3                                     Calle Uruguay 16   
4                            Calle Fermin Caballero 66   
..                                                 ...   
404                Calle Ildefonso Gonzalez Valencia 1   
405  Avenida México 4 Parque De El Retiro. Entrada ...   
406                               Calle Cea Bermudez 1   
407                                Calle Jorge Juan 99   
408                            Calle Via Carpetana 141   

     Distance from station in mts  
0                         1766.07  
1                         9554.24  
2                          723.70  
3                          506.84  
4                         2474.47  
..                            ...  
404                       2798.02  
405                        222.61  
406                        137.49  
407                        222.41  
408                       2332.78  

[409 rows x 3 columns]

In [21]:
# Pruebas
min_bicimad_df[min_bicimad_df['Place of interest'] == 'Colegio Oficial de Arquitectos de Madrid (COAM)']

Place of interest       Place address  \
117  Colegio Oficial de Arquitectos de Madrid (COAM)  Calle Hortaleza 63   

     Distance from station in mts  
117                          8.89

In [22]:
# Creamos un nuevo df bicimad_project_df, haciendo el merge de bicimad con el de los sitios de interés
bicimad_project_df = min_bicimad_df.merge(bicimad_project_df, on=['Place of interest', 'Place address', 'Distance from station in mts'], how='inner')

In [23]:
bicimad_project_df

Place of interest  \
0    A (mas) ECU control urbanístico, S.L. Entidad ...   
1    APFMPS, Residencia, centro de día y centro ocu...   
2    ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
3                   Administración Ciudad Lineal. AEAT   
4                Administración Fermín Caballero. AEAT   
..                                                 ...   
404                            Tanatorio Sur de Madrid   
405   Teatro Municipal de Títeres. Parque de El Retiro   
406                                  Teatros del Canal   
407                                      WiZink Center   
408  Yacimiento paleontológico. Estación de Metro C...   

                                         Place address  \
0               Calle Añastro 11 Planta Baja Puerta 10   
1                        Carretera Colmenar Viejo 13.6   
2                               Camino Valderribas 115   
3                                     Calle Uruguay 16   
4                            Calle Fermin Caballero 66   
..                                                 ...   
404                Calle Ildefonso Gonzalez Valencia 1   
405  Avenida México 4 Parque De El Retiro. Entrada ...   
406                               Calle Cea Bermudez 1   
407                                Calle Jorge Juan 99   
408                            Calle Via Carpetana 141   

     Distance from station in mts                 BiciMAD station  \
0                         1766.07                   254 - Pío XII   
1                         9554.24                217 - Pedro Rico   
2                          723.70           226 - Sierra Toledana   
3                          506.84                  160 - Colombia   
4                         2474.47                217 - Pedro Rico   
..                            ...                             ...   
404                       2798.02                228 - San Delfín   
405                        222.61  60 - Plaza de la Independencia   
406                        137.49                     201 - Canal   
407                        222.41                  62 - O'Donnell   
408                       2332.78                  224 - Caramuel   

                       BiciMAD address  
0                 Avenida Pío XII nº 3  
1                Calle Pedro Rico nº 4  
2          Camino de Valderribas nº 94  
3                  Calle Colombia nº 7  
4                Calle Pedro Rico nº 4  
..                                 ...  
404              Calle San Delfín nº 9  
405     Plaza de la Independencia nº 6  
406  Calle José Abascal frente al nº 2  
407              Calle O'Donnell nº 50  
408               Calle Caramuel nº 51  

[409 rows x 5 columns]

In [24]:
# Incluir columna Type of place (*) 'Instalaciones accesibles NO municipales'

bicimad_project_df.insert(loc=1, column='Type of place', value='Instalaciones accesibles NO municipales')

# df.insert(loc=1, column='NuevaColumna', value=[1, 2, 3, 4]) => https://docs.kanaries.net/es/topics/Pandas/pandas-add-column

# Bicimad Dataframe Final

In [25]:
bicimad_project_df

Place of interest  \
0    A (mas) ECU control urbanístico, S.L. Entidad ...   
1    APFMPS, Residencia, centro de día y centro ocu...   
2    ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
3                   Administración Ciudad Lineal. AEAT   
4                Administración Fermín Caballero. AEAT   
..                                                 ...   
404                            Tanatorio Sur de Madrid   
405   Teatro Municipal de Títeres. Parque de El Retiro   
406                                  Teatros del Canal   
407                                      WiZink Center   
408  Yacimiento paleontológico. Estación de Metro C...   

                               Type of place  \
0    Instalaciones accesibles NO municipales   
1    Instalaciones accesibles NO municipales   
2    Instalaciones accesibles NO municipales   
3    Instalaciones accesibles NO municipales   
4    Instalaciones accesibles NO municipales   
..                                       ...   
404  Instalaciones accesibles NO municipales   
405  Instalaciones accesibles NO municipales   
406  Instalaciones accesibles NO municipales   
407  Instalaciones accesibles NO municipales   
408  Instalaciones accesibles NO municipales   

                                         Place address  \
0               Calle Añastro 11 Planta Baja Puerta 10   
1                        Carretera Colmenar Viejo 13.6   
2                               Camino Valderribas 115   
3                                     Calle Uruguay 16   
4                            Calle Fermin Caballero 66   
..                                                 ...   
404                Calle Ildefonso Gonzalez Valencia 1   
405  Avenida México 4 Parque De El Retiro. Entrada ...   
406                               Calle Cea Bermudez 1   
407                                Calle Jorge Juan 99   
408                            Calle Via Carpetana 141   

     Distance from station in mts                 BiciMAD station  \
0                         1766.07                   254 - Pío XII   
1                         9554.24                217 - Pedro Rico   
2                          723.70           226 - Sierra Toledana   
3                          506.84                  160 - Colombia   
4                         2474.47                217 - Pedro Rico   
..                            ...                             ...   
404                       2798.02                228 - San Delfín   
405                        222.61  60 - Plaza de la Independencia   
406                        137.49                     201 - Canal   
407                        222.41                  62 - O'Donnell   
408                       2332.78                  224 - Caramuel   

                       BiciMAD address  
0                 Avenida Pío XII nº 3  
1                Calle Pedro Rico nº 4  
2          Camino de Valderribas nº 94  
3                  Calle Colombia nº 7  
4                Calle Pedro Rico nº 4  
..                                 ...  
404              Calle San Delfín nº 9  
405     Plaza de la Independencia nº 6  
406  Calle José Abascal frente al nº 2  
407              Calle O'Donnell nº 50  
408               Calle Caramuel nº 51  

[409 rows x 6 columns]

### Exportando .csv BiciMAD_Final

In [45]:
from pathlib import Path
filepath = Path('../data/results/bicimad_project_df.csv', index=False)  
filepath.parent.mkdir(parents=True, exist_ok=True) 
bicimad_project_df.to_csv(filepath, index=False, encoding='utf-8')

#### 2. BiciPARKS

In [27]:
min_biciparks_df = parks_project_df.groupby('Place of interest', as_index=False).agg({'Place address': 'first', 'Distance from parking in mts': 'min'})
min_biciparks_df

Place of interest  \
0    A (mas) ECU control urbanístico, S.L. Entidad ...   
1    APFMPS, Residencia, centro de día y centro ocu...   
2    ASPAYM Madrid (Asociación de Parapléjicos y Pe...   
3                   Administración Ciudad Lineal. AEAT   
4                Administración Fermín Caballero. AEAT   
..                                                 ...   
404                            Tanatorio Sur de Madrid   
405   Teatro Municipal de Títeres. Parque de El Retiro   
406                                  Teatros del Canal   
407                                      WiZink Center   
408  Yacimiento paleontológico. Estación de Metro C...   

                                         Place address  \
0               Calle Añastro 11 Planta Baja Puerta 10   
1                        Carretera Colmenar Viejo 13.6   
2                               Camino Valderribas 115   
3                                     Calle Uruguay 16   
4                            Calle Fermin Caballero 66   
..                                                 ...   
404                Calle Ildefonso Gonzalez Valencia 1   
405  Avenida México 4 Parque De El Retiro. Entrada ...   
406                               Calle Cea Bermudez 1   
407                                Calle Jorge Juan 99   
408                            Calle Via Carpetana 141   

     Distance from parking in mts  
0                         1815.40  
1                         7561.77  
2                         4678.14  
3                         2079.97  
4                         2143.86  
..                            ...  
404                       6064.92  
405                        466.59  
406                       1635.89  
407                       1112.09  
408                       4835.42  

[409 rows x 3 columns]

In [28]:
# Creamos un nuevo df parks_project_df, haciendo el merge de bicimad con el de los sitios de interés
parks_project_dfparks_project_df = min_biciparks_df.merge(parks_project_df, on=['Place of interest', 'Place address', 'Distance from parking in mts'], how='inner')

In [29]:
parks_project_df

Place of interest  \
826                      Administración Montalbán. AEAT   
2906                    Biblioteca del Tribunal Supremo   
1902  Oficina de Correos. Oficina Principal. Paseo d...   
1084  Oficina de Correos. Oficina Principal. Paseo d...   
2902  Biblioteca del Consejo General del Poder judicial   
...                                                 ...   
3992      Oficina de Correos. Sucursal 59. Calle Alanis   
3758  Centro de día y centro de rehabilitación Funda...   
3996  Oficina de Correos. Sucursal 64. Aeropuerto T1...   
3692         Aeropuerto Adolfo Suarez Madrid -  Barajas   
3901  Intercambiador de transporte. Aeropuerto T1 - ...   

                                   Place address         BiciPARK station  \
826                            Calle Montalban 6       Bicipark Montalban   
2906                      Plaza Villa De Paris 1  Bicipark Villa de Paris   
1902                               Paseo Prado 1       Bicipark Recoletos   
1084                               Paseo Prado 1       Bicipark Montalban   
2902              Calle Marques De La Ensenada 8  Bicipark Villa de Paris   
...                                          ...                      ...   
3992                              Calle Alanis 1        Bicipark Aviación   
3758  Plaza Hermanos Falco Y Alvarez De Toledo 1        Bicipark Aviación   
3996      Aeropuerto T1 T2 T3  Sala 1 (Llegadas)        Bicipark Aviación   
3692                          Avenida Hispanidad        Bicipark Aviación   
3901                       Avenida Hispanidad 44        Bicipark Aviación   

               BiciPARK address  Distance from parking in mts  
826       Calle de Montalbán, 5                         29.74  
2906  Calle de Orellana, 23, 21                        105.83  
1902      Paseo de Recoletos, 2                        204.91  
1084      Calle de Montalbán, 5                        205.33  
2902  Calle de Orellana, 23, 21                        210.18  
...                         ...                           ...  
3992  Parking Aviación Española                      24424.29  
3758  Parking Aviación Española                      25032.79  
3996  Parking Aviación Española                      25420.08  
3692  Parking Aviación Española                      25673.22  
3901  Parking Aviación Española                      25673.64  

[4090 rows x 5 columns]

In [30]:
parks_project_df.insert(loc=1, column='Type of place', value='Instalaciones accesibles NO municipales')

# Biciparks Dataframe Final

In [31]:
parks_project_df

Place of interest  \
826                      Administración Montalbán. AEAT   
2906                    Biblioteca del Tribunal Supremo   
1902  Oficina de Correos. Oficina Principal. Paseo d...   
1084  Oficina de Correos. Oficina Principal. Paseo d...   
2902  Biblioteca del Consejo General del Poder judicial   
...                                                 ...   
3992      Oficina de Correos. Sucursal 59. Calle Alanis   
3758  Centro de día y centro de rehabilitación Funda...   
3996  Oficina de Correos. Sucursal 64. Aeropuerto T1...   
3692         Aeropuerto Adolfo Suarez Madrid -  Barajas   
3901  Intercambiador de transporte. Aeropuerto T1 - ...   

                                Type of place  \
826   Instalaciones accesibles NO municipales   
2906  Instalaciones accesibles NO municipales   
1902  Instalaciones accesibles NO municipales   
1084  Instalaciones accesibles NO municipales   
2902  Instalaciones accesibles NO municipales   
...                                       ...   
3992  Instalaciones accesibles NO municipales   
3758  Instalaciones accesibles NO municipales   
3996  Instalaciones accesibles NO municipales   
3692  Instalaciones accesibles NO municipales   
3901  Instalaciones accesibles NO municipales   

                                   Place address         BiciPARK station  \
826                            Calle Montalban 6       Bicipark Montalban   
2906                      Plaza Villa De Paris 1  Bicipark Villa de Paris   
1902                               Paseo Prado 1       Bicipark Recoletos   
1084                               Paseo Prado 1       Bicipark Montalban   
2902              Calle Marques De La Ensenada 8  Bicipark Villa de Paris   
...                                          ...                      ...   
3992                              Calle Alanis 1        Bicipark Aviación   
3758  Plaza Hermanos Falco Y Alvarez De Toledo 1        Bicipark Aviación   
3996      Aeropuerto T1 T2 T3  Sala 1 (Llegadas)        Bicipark Aviación   
3692                          Avenida Hispanidad        Bicipark Aviación   
3901                       Avenida Hispanidad 44        Bicipark Aviación   

               BiciPARK address  Distance from parking in mts  
826       Calle de Montalbán, 5                         29.74  
2906  Calle de Orellana, 23, 21                        105.83  
1902      Paseo de Recoletos, 2                        204.91  
1084      Calle de Montalbán, 5                        205.33  
2902  Calle de Orellana, 23, 21                        210.18  
...                         ...                           ...  
3992  Parking Aviación Española                      24424.29  
3758  Parking Aviación Española                      25032.79  
3996  Parking Aviación Española                      25420.08  
3692  Parking Aviación Española                      25673.22  
3901  Parking Aviación Española                      25673.64  

[4090 rows x 6 columns]

### Exportando .csv Bicipark_Final

In [46]:
from pathlib import Path
filepath = Path('../data/results/biciparks_project_df.csv', index=False)  
filepath.parent.mkdir(parents=True, exist_ok=True) 
parks_project_df.to_csv(filepath, index=False, encoding='utf-8')

## Objetivo: Teniendo los 2 dataframes:

It must provide, **at least,** two options for the final user to select when executing using argparse:
- (1) To get the table for every 'Place of interest' included in the dataset (or a set of them)
- (2) To get the table for a specific 'Place of interest' imputed by the user.

**Fuentes**
- [Usar Argparse en Python](https://www.youtube.com/watch?v=ng-RYmDyE5A&ab_channel=episuarez)

# Reciclaje de códigos

### Exportamos el .csv final de bicimad_project_df y parks_project_df

In [33]:
# from pathlib import Path
# filepath = Path('./modules/bicimad_project_df.csv')  
# filepath.parent.mkdir(parents=True, exist_ok=True) 
# bicimad_project_df.to_csv(filepath)

In [34]:
# from pathlib import Path
# filepath = Path('./modules/parks_project_df.csv')  
# filepath.parent.mkdir(parents=True, exist_ok=True) 
# parks_project_df.to_csv(filepath)

### Argparse

[argparse — Analizador sintáctico (Parser) para las opciones, argumentos y sub-comandos de la línea de comandos](https://docs.python.org/es/3/library/argparse.html)

In [35]:
# from argparse import ArgumentParser

# parser = ArgumentParser(description="'get the table for every 'Place of interest' included in.')
# parser.add_argument('integers', metavar='N', type=int, nargs='+',
#                     help='an integer for the accumulator')
# parser.add_argument('--sum', dest='accumulate', action='store_const',
#                     const=sum, default=max,
#                     help='sum the integers (default: find the max)')

# args = parser.parse_args()
# print(args.accumulate(args.integers))

In [36]:
# import argparse

# parser = argparse.ArgumentParser(description='Process some integers.')
# parser.add_argument('integers', metavar='N', type=int, nargs='+',
#                     help='an integer for the accumulator')
# parser.add_argument('--sum', dest='accumulate', action='store_const',
#                     const=sum, default=max,
#                     help='sum the integers (default: find the max)')

# args = parser.parse_args()
# print(args.accumulate(args.integers))

`module geo_calculations.py.`
[Geopandas library](https://geopandas.org/)

In [37]:
# bicimad_df = pd.read_csv('./data/bicimad_stations.csv')
# bicimad_df.head()

# biciparks_df = pd.read_csv('./data/biciparks_df_v2.csv')
# inst_no_munic_d
# f = pd.read_csv('./data/inst_no_munic_df_v2.csv')

In [38]:
# bicimad_df = bicimad_df.drop(columns='Unnamed: 0')
# bicimad_df = bicimad_df.rename(columns={"mad_long": "BiciMAD Long.",
#                                         "mad_lat" : "BiciMAD Lat."})
# bicimad_df

In [39]:
# biciparks_df = biciparks_df.drop(columns='Unnamed: 0')
# biciparks_df = biciparks_df.drop(columns='zip_code')
# biciparks_df = biciparks_df.rename(columns={"park_long": "BiciPARK Long.",
#                                             "park_lat" : "BiciPARK Lat."})
# biciparks_df

In [40]:
# inst_no_munic_df = inst_no_munic_df.drop(columns='Unnamed: 0')
# inst_no_munic_df = inst_no_munic_df.drop(columns='Postal code')
# inst_no_munic_df = inst_no_munic_df.rename(columns={"place_lat": "Place LAT.",
#                                             "place_long" : "Place LONG."})
# # inst_no_munic_df

In [41]:
# https://www.geeksforgeeks.org/python-program-to-perform-cross-join-in-pandas/

# project_df['distance_meters'] = round(project_df.apply(lambda row: distance_meters(row['BiciMAD Lat.'], row['BiciMAD Long.'], row['Place LAT.'], row['Place LONG.']), axis=1), 2)
# project_df.sort_values(by='distance_meters')

In [42]:
# bicimad_project_df.groupby('Place of interest', as_index=False).agg('Place address',{'distance_meters': ['mean', 'min', 'max']})
# bicimad_project_df.groupby('Place of interest', as_index=False).agg({'Place address': 'first', 'distance_meters': ['mean', 'min', 'max']})

In [43]:
# inst_no_munic_df['Place address'] = inst_no_munic_df['Place address'].apply(lambda row: row.replace('&amp;Ntilde;','Ñ'))

# sitios_interes = bicistation_df['Place of interest'].apply(lambda row: row.replace('&amp;Ntilde;','Ñ')).tolist()

In [44]:
# sitios_interes = [row for row in bicistation_df['Place of interest'] if args.lugar in row].tolist()
# sitios_interes = [row for row in bicistation_df['Place of interest'] if args.lugar in row].tolist()

**Octavio:**

Crear una app de usuario en silla de ruedas que necesita ir a un sitio.

Meterle el lugar de destino y ver si está adaptado a acceso de movilidad reducida.